stan/math/opencl/prim/normal_id_glm_lpdf.hpp

```diff
@@ -47,97 +52,83 @@ namespace math {
-                     y_cl.size());
+                     size(y));
   }
   check_consistent_size(function, "Weight vector", beta, M);
   if (is_sigma_vector) {
     check_size_match(function, "Rows of ", "x_cl", N, "size of ", "sigma",
-                     stan::math::size(sigma));
+                     size(sigma));
   }
   if (is_alpha_vector) {
     check_size_match(function, "Rows of ", "x_cl", N, "size of ", "alpha",
-                     stan::math::size(alpha));
+                     size(alpha));
   }
-  T_sigma_ref sigma_ref = sigma;
-  const auto& sigma_val = value_of_rec(sigma_ref);
-  const auto& sigma_val_vec = to_ref(as_column_vector_or_scalar(sigma_val));
-  check_positive_finite(function, "Scale vector", sigma_val_vec);
-
   if (!include_summand<propto, T_alpha, T_beta, T_scale>::value) {
     return 0;
   }
   if (N == 0) {
     return 0;
   }
 
-  T_beta_ref beta_ref = beta;
-  T_alpha_ref alpha_ref = alpha;
-
-  const auto& beta_val = value_of_rec(beta_ref);
-  const auto& alpha_val = value_of_rec(alpha_ref);
-
-  const auto& beta_val_vec = as_column_vector_or_scalar(beta_val);
-  const auto& alpha_val_vec = as_column_vector_or_scalar(alpha_val);
-
-  T_scale_val inv_sigma = 1 / as_array_or_scalar(sigma_val_vec);
-  Matrix<T_partials_return, Dynamic, 1> y_minus_mu_over_sigma_mat(N);
-  auto y_scaled = y_minus_mu_over_sigma_mat.array();
+  const auto& y_val = value_of(y);
+  const auto& x_val = value_of(x);
+  const auto& alpha_val = value_of(alpha);
+  const auto& beta_val = value_of(beta);
+  const auto& sigma_val = value_of(sigma);
 
-  matrix_cl<double> beta_cl(beta_val_vec);
-  matrix_cl<double> alpha_cl(alpha_val_vec);
-  matrix_cl<double> sigma_cl(sigma_val_vec);
-
-  auto sigma_bc_expr = broadcast<!is_sigma_vector, false>(sigma_cl);
-  auto inv_sigma_expr = elt_divide(1., sigma_bc_expr);
+  auto inv_sigma_expr = elt_divide(1., sigma_val);
   auto y_scaled_expr = elt_multiply(
-      (colwise_optional_broadcast(y_cl) - matrix_vector_multiply(x_cl, beta_cl)
-       - broadcast<!is_alpha_vector, false>(alpha_cl)),
+      (y_val - matrix_vector_multiply(x_val, beta_val) - alpha_val),
       inv_sigma_expr);
   auto mu_derivative_expr = elt_multiply(y_scaled_expr, inv_sigma_expr);
   auto mu_derivative_sum_expr = colwise_sum(mu_derivative_expr);
   auto y_scaled_sq_expr = elt_multiply(y_scaled_expr, y_scaled_expr);
   auto y_scaled_sq_sum_expr = colwise_sum(y_scaled_sq_expr);
   auto sigma_derivative_expr
       = elt_multiply((y_scaled_sq_expr - 1), inv_sigma_expr);
-  auto log_sigma_sum_expr = colwise_sum(log(sigma_bc_expr));
+  auto log_sigma_sum_expr = colwise_sum(log(sigma_val));
 
   const int wgs = y_scaled_sq_sum_expr.rows();
 
-  constexpr bool need_mu_derivative = !is_constant_all<T_beta, T_alpha>::value;
+  constexpr bool need_mu_derivative
+      = !is_constant<T_beta>::value
+        || (!is_constant<T_alpha>::value && is_alpha_vector)
+        || (!is_constant<T_y>::value && is_y_vector);
```

Comment:
[side note] It takes up too much brain ram for me when I see all the `!value`'s chained together. We should add something like `is_differentiable` or something like that

Reply:
Just be ready when adding that - this is used in almost every distribution at least 5 times :)

Reply:
lol yeah I'll put it on my to do eventually list